pkg/controller/worker/reconciler_delete.go

```diff
@@ -0,0 +1,219 @@
+		return fmt.Errorf("Marking all machines for forceful deletion failed: %s", err.Error())
+	}
+
+	// Get the list of all existing machine deployments.
+	existingMachineDeployments := &machinev1alpha1.MachineDeploymentList{}
+	if err := r.client.List(ctx, &client.ListOptions{Namespace: worker.Namespace}, existingMachineDeployments); err != nil {
+		return err
+	}
+
+	// Delete all machine deployments.
+	if err := r.cleanupMachineDeployments(ctx, existingMachineDeployments, nil); err != nil {
+		return fmt.Errorf("Cleaning up machine deployments failed: %s", err.Error())
+	}
+
+	// Delete all machine classes.
+	if err := r.cleanupMachineClasses(ctx, worker.Namespace, workerContext.MachineClassList(), nil); err != nil {
+		return fmt.Errorf("Cleaning up machine classes failed: %s", err.Error())
+	}
+
+	// Delete all machine class secrets.
+	if err := r.cleanupMachineClassSecrets(ctx, worker.Namespace, nil); err != nil {
+		return fmt.Errorf("Cleaning up machine class secrets failed: %s", err.Error())
+	}
+
+	// Wait until all machine resources have been properly deleted.
+	if err := r.waitUntilMachineResourcesDeleted(ctx, worker, workerContext); err != nil {
+		return gardencorev1alpha1helper.DetermineError(fmt.Sprintf("Failed while waiting for all machine resources to be deleted: '%s'", err.Error()))
+	}
+
+	// Delete the machine-controller-manager.
+	r.logger.Info("Deleting the machine-controller-manager", "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
+	if err := workerContext.DeleteMachineControllerManager(ctx); err != nil {
+		return fmt.Errorf("Cleaning up machine-controller-manager failed: %s", err.Error())
+	}
+
+	return nil
+}
+
+// Mark all existing machines to become forcefully deleted.
+func (r *reconciler) markAllMachinesForcefulDeletion(ctx context.Context, namespace string) error {
+	// Mark all existing machines to become forcefully deleted.
+	existingMachines := &machinev1alpha1.MachineList{}
+	if err := r.client.List(ctx, &client.ListOptions{Namespace: namespace}, existingMachines); err != nil {
+		return err
+	}
+
+	var (
+		errorList []error
+		wg        sync.WaitGroup
+	)
+
+	for _, machine := range existingMachines.Items {
+		wg.Add(1)
+		go func(machine *machinev1alpha1.Machine) {
+			defer wg.Done()
+			if err := r.markMachineForcefulDeletion(ctx, machine); err != nil {
+				errorList = append(errorList, err)
+			}
+		}(&machine)
+	}
+
+	wg.Wait()
+	if len(errorList) > 0 {
+		return fmt.Errorf("Labelling machines (to become forcefully deleted) failed: %v", errorList)
+	}
+
+	return nil
+}
+
+// markMachineForcefulDeletion labels a machine object to become forcefully deleted.
+func (r *reconciler) markMachineForcefulDeletion(ctx context.Context, machine *machinev1alpha1.Machine) error {
+	if machine.Labels == nil {
+		machine.Labels = map[string]string{}
+	}
+
+	if val, ok := machine.Labels[forceDeletionLabelKey]; ok && val == forceDeletionLabelValue {
```

Comment:
The label should be public in the machine-controller-manager repository (because it's part of their API), not in ours. Unfortunately, it's not exposed there.